# 1 - Intro

- Import the pandas library as pd
- Importer la bibliothèque pandas en tant que pd

In [ ]:
import pandas as pd

- Import the "titanic.csv" DataFrame in a variable called df
- Importer le DataFrame "titanic.csv" dans une variable nomée df

In [ ]:
df = pd.read_csv("titanic.csv")

- Print the first ten rows of the dataframe
- Afficher les dix premières lignes du dataframe

In [ ]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1       0.0     3.0   
1            2       1.0     1.0   
2            3       1.0     3.0   
3            4       1.0     1.0   
4            5       0.0     3.0   
5            6       0.0     3.0   
6            7       0.0     1.0   
7            8       0.0     3.0   
8            9       1.0     3.0   
9           10       1.0     2.0   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0    1.0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                             Heikkinen, Miss. Laina  female  26.0    0.0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                           Allen, Mr. William Henry    male  35.0    0.0   
5                                   Moran, Mr. James    male   NaN    0.0   
6                            McCarthy, Mr. Timothy J    male  54.0    0.0   
7                     Palsson, Master. Gosta Leonard    male   2.0    3.0   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0    0.0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0    1.0   

   Parch            Ticket     Fare Cabin Embarked  
0    0.0         A/5 21171   7.2500   NaN        S  
1    0.0          PC 17599  71.2833   C85        C  
2    0.0  STON/O2. 3101282   7.9250   NaN        S  
3    0.0            113803  53.1000  C123        S  
4    0.0            373450   8.0500   NaN        S  
5    0.0            330877   8.4583   NaN        Q  
6    0.0             17463  51.8625   E46        S  
7    1.0            349909  21.0750   NaN        S  
8    2.0            347742  11.1333   NaN        S  
9    0.0            237736  30.0708   NaN        C

- Drop the "Cabin" column from the dataframe
- Supprimer la colonne "Cabin" du dataframe

In [ ]:
df= df.drop("Cabin", axis=1)

- Print again the first ten rows of the dataframe
- Afficher à nouveau les dix premières lignes du dataframe

In [ ]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1       0.0     3.0   
1            2       1.0     1.0   
2            3       1.0     3.0   
3            4       1.0     1.0   
4            5       0.0     3.0   
5            6       0.0     3.0   
6            7       0.0     1.0   
7            8       0.0     3.0   
8            9       1.0     3.0   
9           10       1.0     2.0   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0    1.0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                             Heikkinen, Miss. Laina  female  26.0    0.0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                           Allen, Mr. William Henry    male  35.0    0.0   
5                                   Moran, Mr. James    male   NaN    0.0   
6                            McCarthy, Mr. Timothy J    male  54.0    0.0   
7                     Palsson, Master. Gosta Leonard    male   2.0    3.0   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0    0.0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0    1.0   

   Parch            Ticket     Fare Embarked  
0    0.0         A/5 21171   7.2500        S  
1    0.0          PC 17599  71.2833        C  
2    0.0  STON/O2. 3101282   7.9250        S  
3    0.0            113803  53.1000        S  
4    0.0            373450   8.0500        S  
5    0.0            330877   8.4583        Q  
6    0.0             17463  51.8625        S  
7    1.0            349909  21.0750        S  
8    2.0            347742  11.1333        S  
9    0.0            237736  30.0708        C

- Print the description and the info relative to the dataframe
- Afficher la description et les infos relatives au dataframe

In [ ]:
df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  887.000000  887.000000  704.000000  889.000000   
mean    446.000000    0.385569    2.310034   29.722543    0.523060   
std     257.353842    0.487004    0.836350   14.563191    1.103729   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.875000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  887.000000  886.000000  
mean     0.378805   32.169229  
std      0.803998   49.749721  
min      0.000000    0.000000  
25%      0.000000    7.895800  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 11 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     887 non-null    float64
 2   Pclass       887 non-null    float64
 3   Name         884 non-null    object 
 4   Sex          887 non-null    object 
 5   Age          704 non-null    float64
 6   SibSp        889 non-null    float64
 7   Parch        887 non-null    float64
 8   Ticket       883 non-null    object 
 9   Fare         886 non-null    float64
 10  Embarked     889 non-null    object 
dtypes: float64(6), int64(1), object(4)
memory usage: 76.7+ KB


- What can you say about the non-null data and the data type of each column?
- Quelles remarques pouvez-vous faire par rapport au données non-nulles et au type de donnée de chaque colonne?

**Your answer :** ...

**Votre réponse :** on remarque que survived a 4 valeurs null et name est de type object donc on doit le transformer en valeur numerique


- Print again the first 10 rows of the dataframe, then print one categorical feature and one numerical.
- Afficher de nouveau les 10 premières lignes du dataframe, puis afficher un attribut catégorique et un autre numérique.

In [ ]:
df.head(10)

PassengerId  Survived  Pclass  \
0            1       0.0     3.0   
1            2       1.0     1.0   
2            3       1.0     3.0   
3            4       1.0     1.0   
4            5       0.0     3.0   
5            6       0.0     3.0   
6            7       0.0     1.0   
7            8       0.0     3.0   
8            9       1.0     3.0   
9           10       1.0     2.0   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0    1.0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                             Heikkinen, Miss. Laina  female  26.0    0.0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                           Allen, Mr. William Henry    male  35.0    0.0   
5                                   Moran, Mr. James    male   NaN    0.0   
6                            McCarthy, Mr. Timothy J    male  54.0    0.0   
7                     Palsson, Master. Gosta Leonard    male   2.0    3.0   
8  Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)  female  27.0    0.0   
9                Nasser, Mrs. Nicholas (Adele Achem)  female  14.0    1.0   

   Parch            Ticket     Fare Embarked  
0    0.0         A/5 21171   7.2500        S  
1    0.0          PC 17599  71.2833        C  
2    0.0  STON/O2. 3101282   7.9250        S  
3    0.0            113803  53.1000        S  
4    0.0            373450   8.0500        S  
5    0.0            330877   8.4583        Q  
6    0.0             17463  51.8625        S  
7    1.0            349909  21.0750        S  
8    2.0            347742  11.1333        S  
9    0.0            237736  30.0708        C

In [ ]:
df[["Age","Name"]]

Age                                               Name
0    22.0                            Braund, Mr. Owen Harris
1    38.0  Cumings, Mrs. John Bradley (Florence Briggs Th...
2    26.0                             Heikkinen, Miss. Laina
3    35.0       Futrelle, Mrs. Jacques Heath (Lily May Peel)
4    35.0                           Allen, Mr. William Henry
..    ...                                                ...
886  27.0                              Montvila, Rev. Juozas
887  19.0                       Graham, Miss. Margaret Edith
888   NaN           Johnston, Miss. Catherine Helen "Carrie"
889  26.0                              Behr, Mr. Karl Howell
890  32.0                                Dooley, Mr. Patrick

[891 rows x 2 columns]

# 2- Cleaning

- Check if any missing values are in the dataframe, if there is, print how many.
- Vérifiez s'il y a des valeurs manquantes dans le dataframe, s'il y en a, afficher combien.

In [ ]:
df.isnull()

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket  \
0          False     False   False  False  False  False  False  False   False   
1          False     False   False  False  False  False  False  False   False   
2          False     False   False  False  False  False  False  False   False   
3          False     False   False  False  False  False  False  False   False   
4          False     False   False  False  False  False  False  False   False   
..           ...       ...     ...    ...    ...    ...    ...    ...     ...   
886        False     False   False  False  False  False  False  False   False   
887        False     False   False  False  False  False  False  False   False   
888        False     False   False  False  False   True  False  False   False   
889        False     False   False  False  False  False  False  False   False   
890        False     False   False  False  False  False  False  False   False   

      Fare  Embarked  
0    False     False  
1    False     False  
2    False     False  
3    False     False  
4    False     False  
..     ...       ...  
886  False     False  
887  False     False  
888  False     False  
889  False     False  
890  False     False  

[891 rows x 11 columns]

In [ ]:
df.isnull().sum()

Unnamed: 0           0
make                 0
fuel_type            0
aspiration           0
num_doors            0
body_style           0
drive_wheels         0
engine_location      0
wheel_base           0
length               0
width                0
height               0
curb_weight          0
engine_type          0
num_cylinders        0
engine_size          0
bore                 0
stroke               0
compression_ratio    0
horsepower           0
peak_rpm             0
price                0
dtype: int64

- Print the total number of missing values in the dataframe.
- Afficher le nombre totale de valeurs manquantes dans le dataframe.

In [ ]:
df.isnull().sum().sum()

227

---

- By the end of this section, this number must be 0 (which means no missing values left)
- À la fin de cette section, ce nombre doit être égale a 0 (ce qui signifie qu'il ne reste plus aucune valeur manquante)

-----

- First create a copy of this dataframe (df_copy = df)
- Créez d'abord une copie de ce dataframe (df_copy = df)

In [ ]:
df_copy= df

- Use the function you learned in the course to delete the missing values in the copy you created from the dataframe
- Utilisez la fonction que vous avez apprise dans le cours pour supprimer les valeurs manquantes dans la copie que vous avez créée à partir du dataframe

In [ ]:
df_copy.dropna(axis=0,how='any',thresh=None, inplace=True)

- Check the new number of missing values.
- Vérifier le nouveau nombre de valeurs manquantes.

In [ ]:
df_copy.isnull().sum().sum()

0

- Check the new length of the dataframe.
- Vérifier la nouvelle taille du dataframe.

In [ ]:
print(len(df_copy))

693


- What did you notice ?
- Qu'avez-vous remarqué ?

**Your answer :** ...

**Votre réponse :on remarque le nombre a diminué de 891 à 693 


- Create a new copy of this dataframe (df_copy2 = df)
- Créez une nouvelle copie de ce dataframe (df_copy2 = df)

In [ ]:
df_copy2= df

- Print the mean of the Age column
- Afficher la moyenne des valeurs de la colonne Age

In [ ]:
df_copy2['Age'].mean()

29.72246753246753

- Print the median of the Age column
- Afficher la valeur médiane des valeurs de la colonne Age

In [ ]:
df_copy2['Age'].median()

28.0

- Print the mode of the Age column
- Afficher le mode des valeurs de la colonne Age

In [ ]:
df_copy2['Age'].mode()

0    24.0
dtype: float64

- Now that you have these three values, chose one and use it to replace the missing values of the age
- Maintenant que vous avez ces trois valeurs, choisissez-en une et utilisez-la pour remplacer les valeurs manquantes de la colonne Age

In [ ]:
df_copy2['Age'].fillna(df_copy2['Age'].mean(), inplace=True)

- Check now the number of missing values of the age column
- Vérifier maintenant le nombre de valeurs manquantes de la colonne Age.

In [ ]:
df_copy2['Age'].isnull().sum()

0

# 3 - Feature transformation

- Import the "auto.csv" dataframe using pandas and print the first 10 rows
- En utilisant pandas, importer le dataframe "auto.csv" et afficher les 10 premieres lignes

In [ ]:
df= pd.read_csv('auto_ds.csv')

In [ ]:
df.head(10)

Unnamed: 0       make fuel_type aspiration num_doors body_style  \
0           0       audi       gas        std      four      sedan   
1           1       audi       gas        std      four      sedan   
2           2       audi       gas        std      four      sedan   
3           3       audi       gas      turbo      four      sedan   
4           4        bmw       gas        std       two      sedan   
5           5        bmw       gas        std      four      sedan   
6           6        bmw       gas        std       two      sedan   
7           7        bmw       gas        std      four      sedan   
8           8  chevrolet       gas        std       two  hatchback   
9           9  chevrolet       gas        std       two  hatchback   

  drive_wheels engine_location  wheel_base  length  ...  curb_weight  \
0          fwd           front        99.8   176.6  ...         2337   
1          4wd           front        99.4   176.6  ...         2824   
2          fwd           front       105.8   192.7  ...         2844   
3          fwd           front       105.8   192.7  ...         3086   
4          rwd           front       101.2   176.8  ...         2395   
5          rwd           front       101.2   176.8  ...         2395   
6          rwd           front       101.2   176.8  ...         2710   
7          rwd           front       101.2   176.8  ...         2765   
8          fwd           front        88.4   141.1  ...         1488   
9          fwd           front        94.5   155.9  ...         1874   

   engine_type  num_cylinders engine_size  bore  stroke  compression_ratio  \
0          ohc           four         109  3.19    3.40               10.0   
1          ohc           five         136  3.19    3.40                8.0   
2          ohc           five         136  3.19    3.40                8.5   
3          ohc           five         131  3.13    3.40                8.3   
4          ohc           four         108  3.50    2.80                8.8   
5          ohc           four         108  3.50    2.80                8.8   
6          ohc            six         164  3.31    3.19                9.0   
7          ohc            six         164  3.31    3.19                9.0   
8            l          three          61  2.91    3.03                9.5   
9          ohc           four          90  3.03    3.11                9.6   

   horsepower  peak_rpm    price  
0       102.0    5500.0  13950.0  
1       115.0    5500.0  17450.0  
2       110.0    5500.0  17710.0  
3       140.0    5500.0  23875.0  
4       101.0    5800.0  16430.0  
5       101.0    5800.0  16925.0  
6       121.0    4250.0  20970.0  
7       121.0    4250.0  21105.0  
8        48.0    5100.0   5151.0  
9        70.0    5400.0   6295.0  

[10 rows x 22 columns]

- Check the datatypes of the dataframe (use .dtypes)
- Vérifier les types de données dans le dataframe (Utiliser .dtypes)

In [ ]:
df.dtypes

Unnamed: 0             int64
make                  object
fuel_type             object
aspiration            object
num_doors             object
body_style            object
drive_wheels          object
engine_location       object
wheel_base           float64
length               float64
width                float64
height               float64
curb_weight            int64
engine_type           object
num_cylinders         object
engine_size            int64
bore                 float64
stroke               float64
compression_ratio    float64
horsepower           float64
peak_rpm             float64
price                float64
dtype: object

- In this section we will be working on the "num_doors" column
- Dans cette section, nous allons travailler sur la colonne "num_doors"

---

- The following instruction will give you the count of each distinct value in the column
- L'instruction suivante vous donnera le nombre de chaque valeur distincte dans la colonne

In [ ]:
df["num_doors"].value_counts()

four    95
two     64
Name: num_doors, dtype: int64

- Create a dictionary to change the values from "four" to 4 and from "two" to 2.
- Créer un dictionnaire pour modifier les valeurs de "four" à 4 et de "two" à 2.

In [ ]:
clean = {"num_doors": {"four":4, "two":2}}


- Use the dictionary to replace categorical values with numerical values
- Utiliser le dictionnaire pour remplacer les valeurs catégorielles par des valeurs numériques

In [ ]:
df.replace(clean, inplace= True)

- Import LabelEncoder from the sklearn.preprocessing library
- Importer LabelEncoder depuis la bibliothèque sklearn.preprocessing

In [ ]:
from sklearn.preprocessing import LabelEncoder

- Create a label encoder variable named *enc*
- Créez une variable label encoder nommée *enc*

In [ ]:
enc = LabelEncoder()

- Use the Label Encoder you created to encode the numeric values of the column "body_style" of the dataframe
- Utiliser le Label Encoder que vous avez créez pour encoder les valeurs de le colonne "body_style" a valeurs numériques restantes du dataframe

In [ ]:
df["body_style"]=enc.fit_transform(df["body_style"]
                                  )

In [ ]:
df['body_style'].values

array([3, 3, 3, 3, 3, 3, 3, 3, 2, 2, 3, 2, 2, 2, 2, 3, 3, 4, 2, 2, 2, 2,
       2, 2, 3, 4, 2, 2, 3, 3, 3, 3, 3, 2, 2, 2, 3, 3, 2, 3, 2, 3, 2, 3,
       3, 4, 1, 3, 0, 2, 2, 2, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 4, 3, 2,
       3, 4, 1, 2, 3, 3, 4, 3, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 2, 2, 2, 3,
       3, 4, 2, 2, 3, 2, 3, 2, 3, 2, 2, 2, 3, 3, 3, 3, 3, 4, 4, 4, 4, 2,
       2, 2, 4, 4, 4, 3, 2, 3, 2, 3, 2, 3, 3, 2, 3, 2, 1, 1, 2, 1, 2, 0,
       3, 3, 2, 3, 2, 2, 2, 3, 3, 3, 3, 3, 3, 3, 3, 2, 3, 4, 3, 4, 3, 4,
       3, 3, 3, 3, 3])

- Now that you know how to encode values categorical variable into a number do the same process for the remaining columns.
- Maintenant que vous savez comment coder les valeurs d'une variable catégorielle dans un nombre, procédez de la même manière pour les colonnes restantes.

Hint : you can do the columns one by one or use the .apply() method to encode all the columns at once

Indice : Vous pouvez faire les colonnes une par une ou utiliser la méthode .apply() pour encoder toutes les colonnes à la fois

In [ ]:
col=["make" ,"fuel_type", "aspiration", "drive_wheels", "engine_location","engine_type", "num_cylinders"]
df[col]=df[col].apply(enc.fit_transform)

In [ ]:
df.dtypes

Unnamed: 0             int64
make                   int64
fuel_type              int64
aspiration             int64
num_doors              int64
body_style             int64
drive_wheels           int64
engine_location        int64
wheel_base           float64
length               float64
width                float64
height               float64
curb_weight            int64
engine_type            int64
num_cylinders          int64
engine_size            int64
bore                 float64
stroke               float64
compression_ratio    float64
horsepower           float64
peak_rpm             float64
price                float64
dtype: object

# 4 - One hot encoding

- Import the "titanic.csv" dataframe
- Importer le dataframe "titanic.csv"

In [ ]:
df = pd.read_csv("titanic.csv")

In [ ]:
df.head()

PassengerId  Survived  Pclass  \
0            1       0.0     3.0   
1            2       1.0     1.0   
2            3       1.0     3.0   
3            4       1.0     1.0   
4            5       0.0     3.0   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0    1.0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0    1.0   
2                             Heikkinen, Miss. Laina  female  26.0    0.0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0    1.0   
4                           Allen, Mr. William Henry    male  35.0    0.0   

   Parch            Ticket     Fare Cabin Embarked  
0    0.0         A/5 21171   7.2500   NaN        S  
1    0.0          PC 17599  71.2833   C85        C  
2    0.0  STON/O2. 3101282   7.9250   NaN        S  
3    0.0            113803  53.1000  C123        S  
4    0.0            373450   8.0500   NaN        S

- Get the one hot encoding of the 'Sex' column
- Obtener le one hot encoding de la colonne 'Sex'

In [ ]:
one_hot = pd.get_dummies(df["Sex"])

In [ ]:
one_hot

female  male
0         0     1
1         1     0
2         1     0
3         1     0
4         0     1
..      ...   ...
886       0     1
887       1     0
888       1     0
889       0     1
890       0     1

[891 rows x 2 columns]

- Drop the 'Sex' column of the dataframe and add the one hot encode result to the dataframe
- Supprimer la colonne 'Sex' du dataframe et ajouter le résultat du one hot encode

Hint : to add columns to a dataframe use .join()

Indice : Pour rajouter des colonnes au dataframe utiliser .join()

In [ ]:
df= df.drop("Sex",axis=1)

In [ ]:
df = df.join(one_hot)

- Print the first 5 rows of the dataframe
- Afficher les 5 premieres ligne du dataframe

In [ ]:
df.head(5)

PassengerId  Survived  Pclass  \
0            1       0.0     3.0   
1            2       1.0     1.0   
2            3       1.0     3.0   
3            4       1.0     1.0   
4            5       0.0     3.0   

                                                Name   Age  SibSp  Parch  \
0                            Braund, Mr. Owen Harris  22.0    1.0    0.0   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  38.0    1.0    0.0   
2                             Heikkinen, Miss. Laina  26.0    0.0    0.0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  35.0    1.0    0.0   
4                           Allen, Mr. William Henry  35.0    0.0    0.0   

             Ticket     Fare Cabin Embarked  female  male  
0         A/5 21171   7.2500   NaN        S       0     1  
1          PC 17599  71.2833   C85        C       1     0  
2  STON/O2. 3101282   7.9250   NaN        S       1     0  
3            113803  53.1000  C123        S       1     0  
4            373450   8.0500   NaN        S       0     1